lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -282,312 +546,656 @@ class OpaqueValueVisitor {
+
+  // It's only use should be dealloc_stacks.
+  for (Operand *use : allocInst->getUses()) {
+    pass.deleter.forceDelete(cast<DeallocStackInst>(use->getUser()));
+  }
+  pass.deleter.forceDelete(allocInst);
+}
+
+// Create alloc_stack that dominates an owned value \p value. Create
+// jointly-postdominating dealloc_stack instructions.  Nesting will be fixed
+// later.
+//
+// Any value that may be used by a return instruction must be deallocated
+// immediately before the return. This allows the return to be rewritten by
+// loading from storage.
+AllocStackInst *OpaqueStorageAllocation::
+createStackAllocation(SILValue value) {
+  assert(value.getOwnershipKind() != OwnershipKind::Guaranteed && 
+         "creating storage for a guaranteed value implies a copy");
+
+#ifndef NDEBUG
+  // Instructions that produce an opened type never reach here because they
+  // have guaranteed ownership--they project their storage. We reach this
+  // point after the opened value has been copied.
+  if (auto *defInst = value->getDefiningInstruction()) {
+    if (auto *singleValue = dyn_cast<SingleValueInstruction>(defInst)) {
+      assert(!cast<SingleValueInstruction>(defInst)->getDefinedOpenedArchetype()
+             && "owned open_existential is unsupported");
     }
   }
+#endif
 
-  SILBuilder allocBuilder(pass.F->begin()->begin());
-  allocBuilder.setSILConventions(
-       SILModuleConventions::getLoweredAddressConventions(pass.F->getModule()));
-  AllocStackInst *allocInstr =
-      allocBuilder.createAllocStack(value.getLoc(), value->getType());
+  SILType allocTy = value->getType();
 
-  storage.storageAddress = allocInstr;
+  // For opened existential types, allocate stack space at the type
+  // definition. Allocating as early as possible provides more opportunity for
+  // creating use projections into value.
+  SILInstruction *firstOpeningInst = nullptr;
+  allocTy.getASTType().visit([&](CanType type) {
+    auto archetype = dyn_cast<ArchetypeType>(type);
+    if (!archetype)
+      return;
+
+    if (auto openedTy = getOpenedArchetypeOf(archetype)) {
+      auto openingVal =
+          pass.getModule()->getRootOpenedArchetypeDef(openedTy, pass.function);
 
-  // Insert stack deallocations.
-  for (TermInst *termInst : pass.returnInsts) {
-    SILBuilder deallocBuilder(termInst);
-    deallocBuilder.setSILConventions(
-       SILModuleConventions::getLoweredAddressConventions(pass.F->getModule()));
-    deallocBuilder.createDeallocStack(allocInstr->getLoc(), allocInstr);
+      auto *openingInst = openingVal->getDefiningInstruction();
+      assert(openingVal && "all opened archetypes should be resolved");
+      if (firstOpeningInst
+          && pass.domInfo->dominates(firstOpeningInst, openingInst)) {
+        return;
+      }
+      firstOpeningInst = openingInst;
+    }
+  });
+  auto allocPt = firstOpeningInst ? std::next(firstOpeningInst->getIterator())
+                                  : pass.function->begin()->begin();
```

Comment:
@eeckstein Initially (4 years ago) I was paranoid about -Onone compile time. Not anymore. We certainly could do an analysis here to figure out the lifetime of stack slots.

I don't think shrink wrapping will make any difference until address lowering is run after inlining, so nothing needs to happen in this PR. It is, however, still worth thinking about.

We can't actually figure out the optimal alloc_stack placement until we know all the places where we'll end up storing into the reused location. This would be a completely separate analysis. I can't think of any reason to complicate address lowering with that analysis. If that's a valuable thing to do, then why should it be confined to this pass? And why confined to address-only values? Shouldn't we perform shrink wrapping on all stack allocations?

I'm open to all suggestions.

@aschwaighofer any ideas?

Reply:
My concern is that we have several peephole optimizations which only kick in if all uses of an `alloc_stack` are in the same block.

But I agree, this pass is already very complex. It would make more sense to implement stack-location shrink wrapping as a separate pass. But it might be necessary to do it before opaque values are turned on to not regress performance.

Reply:
> Another concern is that we have several peephole optimizations which only kick in if all uses of an alloc_stack are in the same block.

That's too bad. For the sake of optimizer stability, I don't think we should have any peephole optimizations that make such an assumption.

I agree that shrink-wrapping should be done before address lowering is enabled after Mem2Reg runs at -O. Can you think of any reason this would cause a problem before Mem2Reg?

Note that any address-only local variables will still have the same local alloc_stack scope prior to Mem2Reg.

I'm not sure we would want to do shrink-wrapping at -Onone. We would need to figure out how to preserve variable inspection after last-use.